pkg/reconciler/knativeserving/common/gateway.go

```diff
@@ -23,9 +24,19 @@ func GatewayTransform(instance *servingv1alpha1.KnativeServing, log *zap.Sugared
 }
 
 func updateKnativeIngressGateway(gatewayOverrides servingv1alpha1.IstioGatewayOverride, u *unstructured.Unstructured) error {
-	if len(gatewayOverrides.Selector) > 0 {
+	if len(gatewayOverrides.Selector) > 0 || len(gatewayOverrides.Servers) > 0 {
 		log.Debugw("Updating Gateway", "name", u.GetName(), "gatewayOverrides", gatewayOverrides)
-		unstructured.SetNestedStringMap(u.Object, gatewayOverrides.Selector, "spec", "selector")
+		if len(gatewayOverrides.Selector) > 0 {
+			unstructured.SetNestedStringMap(u.Object, gatewayOverrides.Selector, "spec", "selector")
+		}
+		if len(gatewayOverrides.Servers) > 0 {
```

Comment:
When len(gatewayOverrides.Servers) > 0, we need to focus on gatewayOverrides.Servers only, which mean instead of converting gatewayOverrides, covert gatewayOverrides.Servers, by creating newReferences to hold all the servers, if necessary.

You can check the following section as a reference.
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go#L199

In the end, only set the field servers in spec, like u.setNestedField(newReferences, "spec", "servers").

This will better separate the logic for gatewayOverrides.Selector and gatewayOverrides.Servers.

Reply:
I updated the code.